### ISSUE TITLE
lint generic type argument of a generic type that could be `: ?Sized`

### ISSUE THREAD

**Issue by oli-obk** [Manishearth/rust-clippy#1368]
```rust
struct A<T: Trait>(Box<T>);
```

should lint that one should write `T: Trait + ?Sized`.

Motivation: http://stackoverflow.com/questions/40846232/rust-compiler-does-not-see-structure-as-sized

To reduce false positives, this should only lint if the `impl X for A` and `impl A`, will continue working if they are also changed.
